# genotyping-matrices

Scripts and python modules for genotyping cells from perturbation screens followed by scRNA-seq. Currently compatible with the 10X genomics Chromium V2 chemistry.

Required input: 
* Genotype barcodes amplified from 10X genomics scRNA-seq cDNA: Read1 will be the cell/molecule barcode and Read2 should have the genotype barcode. Read2 needs to be in bam format. You can use PicardTools FastqToBam to convert the Fastq to bam, or align Read2 to the known genotype scaffold as an initial filtering step.
* Counts matrix generated by 10X Genomics CellRanger as a tab separated values (.tsv) file. You can use the convert_chromium_matrix.R script to convert the sparse matrix from CellRanger output to a "dense" tsv matrix.
* Text file mapping barcodes to genotypes. These genotypes can be genes, CRISPR gRNAs, or any other genotype (mutations, etc.)

Usage (Replace filenames in brackets with the names of your files):
1. Tag Read2 genotype reads with cell and molecule barcodes using convert_gRNA.py: `python convert_gRNA.py [read1_cell_molecule_barcodes].fastq [read2_genotype_barcodes].bam`
2. Parse the cell barcodes with parseCellBarcodes.py. `python parseCellBarcodes.py [counts_matrix].tsv [read2_genotypes_barcodes].tagged.bam [barcode_length (integer)] [barcode_upstream_sequence] [barcode_downstream_sequence]`
3. Plot UMI fraction vs Read fraction distribution to set appropriate cutoffs to filter chimeric reads. `python [read2_genotype_barcodes]_cell_barcodes.pickle [min_read_count_fraction]`
4. Generate the genotype to cell dictionaries. 
